src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -3954,47 +3801,416 @@ ColocationGroupDeleteCommand(uint32 colocationId)
 
 
 /*
- * ColocationGroupCreateCommandList returns the full list of commands for syncing
- * pg_dist_colocation.
+ * SetMetadataSyncNodesFromNodeList sets list of nodes that needs to be metadata
+ * synced among given node list into metadataSyncContext.
  */
-List *
-ColocationGroupCreateCommandList(void)
+void
+SetMetadataSyncNodesFromNodeList(MetadataSyncContext *context, List *nodeList)
+{
+	List *activatedWorkerNodeList = NIL;
+
+	WorkerNode *node = NULL;
+	foreach_ptr(node, nodeList)
+	{
+		if (EnableMetadataSync && NodeIsPrimary(node))
+		{
+			/* warn if we have coordinator in nodelist */
+			if (NodeIsCoordinator(node))
+			{
+				ereport(NOTICE, (errmsg("%s:%d is the coordinator and already contains "
+										"metadata, skipping syncing the metadata",
+										node->workerName, node->workerPort)));
+				continue;
+			}
+
+			activatedWorkerNodeList = lappend(activatedWorkerNodeList, node);
+		}
+	}
+
+	context->activatedWorkerNodeList = activatedWorkerNodeList;
+}
+
+
+/*
+ * EstablishAndSetMetadataSyncBareConnections establishes and sets
+ * connections used throughout nontransactional metadata sync.
+ */
+void
+EstablishAndSetMetadataSyncBareConnections(MetadataSyncContext *context)
+{
+	Assert(MetadataSyncTransMode == METADATA_SYNC_NON_TRANSACTIONAL);
+
+	int connectionFlags = 0;
```

Comment:
I'm still thinking whether to use connectionFlags=REQUIRE_METADATA_CONNECTION or not. I think we should do that

Reply:
The main reason I consider this is the following check where ensure we cannot have multiple metadata connections: https://github.com/citusdata/citus/blob/f061dbb253d27bb7f16b0f43665d584a23e25f44/src/backend/distributed/connection/connection_management.c#L531-L536

Reply:
ping on this. Does that break anything?

Reply:
I added REQUIRE_METADATA_CONNECTION flag to access connection.